File: kotlin/engine/src/main/kotlin/org/hpb/engine/nostr/Nip44.kt
package org.hpb.engine.nostr

import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.ChaCha20ParameterSpec
import javax.crypto.spec.SecretKeySpec
import org.hpb.engine.Secp

/**
 * NIP-44 v2 encryption (the protocol's channel for submissions and PSBT
 * envelopes), validated against the official vendored test vectors.
 */
object Nip44 {
    private const val VERSION: Byte = 2
    private const val MIN_LEN = 1
    private const val MAX_LEN = 65535

    private fun hmacSha256(key: ByteArray, vararg chunks: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        chunks.forEach(mac::update)
        return mac.doFinal()
    }

    internal fun hkdfExtract(salt: ByteArray, ikm: ByteArray): ByteArray = hmacSha256(salt, ikm)

    internal fun hkdfExpand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        val out = ArrayList<Byte>(length)
        var block = ByteArray(0)
        var counter = 1
        while (out.size < length) {
            block = hmacSha256(prk, block, info, byteArrayOf(counter.toByte()))
            out.addAll(block.toList())
            counter++
        }
        return out.take(length).toByteArray()
    }

    /** Shared conversation key for (our privkey, their x-only pubkey). */
    fun conversationKey(privkey: ByteArray, peerXonly: ByteArray): ByteArray =
        hkdfExtract("nip44-v2".toByteArray(), Secp.ecdhX(privkey, peerXonly))

    internal data class MessageKeys(
        val chachaKey: ByteArray,
        val chachaNonce: ByteArray,
        val hmacKey: ByteArray,
    )

    internal fun messageKeys(conversationKey: ByteArray, nonce: ByteArray): MessageKeys {
        val expanded = hkdfExpand(conversationKey, nonce, 76)
        return MessageKeys(
            expanded.copyOfRange(0, 32),
            expanded.copyOfRange(32, 44),
            expanded.copyOfRange(44, 76),
        )
    }

    internal fun paddedLength(unpadded: Int): Int {
        require(unpadded >= MIN_LEN) { "empty plaintext" }
        if (unpadded <= 32) return 32
        val nextPower = Integer.highestOneBit(unpadded - 1) shl 1
        val chunk = if (nextPower <= 256) 32 else nextPower / 8
        return chunk * ((unpadded - 1) / chunk + 1)
    }

    private fun pad(plaintext: ByteArray): ByteArray {
        require(plaintext.size in MIN_LEN..MAX_LEN) { "invalid plaintext length" }
        val padded = ByteArray(2 + paddedLength(plaintext.size))
        padded[0] = (plaintext.size ushr 8).toByte()
        padded[1] = plaintext.size.toByte()
        plaintext.copyInto(padded, 2)
        return padded
    }

    private fun unpad(padded: ByteArray): ByteArray {
        val length = ((padded[0].toInt() and 0xff) shl 8) or (padded[1].toInt() and 0xff)
        require(length in MIN_LEN..MAX_LEN && padded.size == 2 + paddedLength(length)) {
            "invalid padding"
        }
        return padded.copyOfRange(2, 2 + length)
    }

    private fun chacha(key: ByteArray, nonce: ByteArray, data: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("ChaCha20")
        cipher.init(
            Cipher.ENCRYPT_MODE, SecretKeySpec(key, "ChaCha20"), ChaCha20ParameterSpec(nonce, 0),
        )
        return cipher.doFinal(data)
    }

    fun encrypt(plaintext: String, conversationKey: ByteArray, nonce: ByteArray): String {
        val keys = messageKeys(conversationKey, nonce)
        val ciphertext = chacha(keys.chachaKey, keys.chachaNonce, pad(plaintext.toByteArray()))
        val mac = hmacSha256(keys.hmacKey, nonce, ciphertext)
        return Base64.getEncoder()
            .encodeToString(byteArrayOf(VERSION) + nonce + ciphertext + mac)
    }

    fun decrypt(payload: String, conversationKey: ByteArray): String {
        require(!payload.startsWith("#")) { "unsupported future version" }
        val raw = Base64.getDecoder().decode(payload)
        require(raw.size >= 1 + 32 + 32 + 34 && raw[0] == VERSION) { "invalid payload" }
        val nonce = raw.copyOfRange(1, 33)
        val ciphertext = raw.copyOfRange(33, raw.size - 32)
        val mac = raw.copyOfRange(raw.size - 32, raw.size)
        val keys = messageKeys(conversationKey, nonce)
        require(hmacSha256(keys.hmacKey, nonce, ciphertext).contentEquals(mac)) { "invalid MAC" }
        return String(unpad(chacha(keys.chachaKey, keys.chachaNonce, ciphertext)))
    }
}

File: ios/HpbCore/Sources/HpbCore/Nip44.swift
import CryptoKit
import Foundation

/// NIP-44 v2 encryption (the protocol's channel for submissions),
/// validated against the official vendored test vectors — the same file
/// the Kotlin reference pins.
public enum Nip44 {
    private static let version: UInt8 = 2
    private static let minLen = 1
    private static let maxLen = 65535

    private static func hmacSha256(_ key: [UInt8], _ chunks: [[UInt8]]) -> [UInt8] {
        var mac = HMAC<CryptoKit.SHA256>(key: SymmetricKey(data: Data(key)))
        for chunk in chunks { mac.update(data: Data(chunk)) }
        return Array(mac.finalize())
    }

    static func hkdfExtract(salt: [UInt8], ikm: [UInt8]) -> [UInt8] {
        hmacSha256(salt, [ikm])
    }

    static func hkdfExpand(prk: [UInt8], info: [UInt8], length: Int) -> [UInt8] {
        var out = [UInt8]()
        var block = [UInt8]()
        var counter: UInt8 = 1
        while out.count < length {
            block = hmacSha256(prk, [block, info, [counter]])
            out.append(contentsOf: block)
            counter &+= 1
        }
        return Array(out.prefix(length))
    }

    /// Shared conversation key for (our privkey, their x-only pubkey).
    public static func conversationKey(_ privkey: [UInt8], _ peerXonly: [UInt8]) throws -> [UInt8] {
        hkdfExtract(salt: Array("nip44-v2".utf8), ikm: try Secp.ecdhX(privkey, peerXonly))
    }

    struct MessageKeys {
        let chachaKey: [UInt8]
        let chachaNonce: [UInt8]
        let hmacKey: [UInt8]
    }

    static func messageKeys(_ conversationKey: [UInt8], _ nonce: [UInt8]) -> MessageKeys {
        let expanded = hkdfExpand(prk: conversationKey, info: nonce, length: 76)
        return MessageKeys(
            chachaKey: Array(expanded[0..<32]),
            chachaNonce: Array(expanded[32..<44]),
            hmacKey: Array(expanded[44..<76])
        )
    }

    static func paddedLength(_ unpadded: Int) throws -> Int {
        guard unpadded >= minLen else { throw HpbError.invalid("empty plaintext") }
        if unpadded <= 32 { return 32 }
        var nextPower = 1
        while nextPower < unpadded { nextPower <<= 1 }
        let chunk = nextPower <= 256 ? 32 : nextPower / 8
        return chunk * ((unpadded - 1) / chunk + 1)
    }

    private static func pad(_ plaintext: [UInt8]) throws -> [UInt8] {
        guard plaintext.count >= minLen, plaintext.count <= maxLen else {
            throw HpbError.invalid("invalid plaintext length")
        }
        var padded = [UInt8](repeating: 0, count: 2 + (try paddedLength(plaintext.count)))
        padded[0] = UInt8(plaintext.count >> 8)
        padded[1] = UInt8(plaintext.count & 0xff)
        for (i, byte) in plaintext.enumerated() { padded[2 + i] = byte }
        return padded
    }

    private static func unpad(_ padded: [UInt8]) throws -> [UInt8] {
        guard padded.count >= 2 else { throw HpbError.invalid("invalid padding") }
        let length = Int(padded[0]) << 8 | Int(padded[1])
        guard length >= minLen, length <= maxLen,
              padded.count == 2 + (try paddedLength(length)) else {
            throw HpbError.invalid("invalid padding")
        }
        return Array(padded[2..<2 + length])
    }

    public static func encrypt(
        _ plaintext: String,
        _ conversationKey: [UInt8],
        _ nonce: [UInt8]
    ) throws -> String {
        let keys = messageKeys(conversationKey, nonce)
        let ciphertext = ChaCha20.xor(
            key: keys.chachaKey, nonce12: keys.chachaNonce,
            data: try pad(Array(plaintext.utf8))
        )
        let mac = hmacSha256(keys.hmacKey, [nonce, ciphertext])
        return Data([version] + nonce + ciphertext + mac).base64EncodedString()
    }

    public static func decrypt(_ payload: String, _ conversationKey: [UInt8]) throws -> String {
        guard !payload.hasPrefix("#") else { throw HpbError.invalid("unsupported future version") }
        guard let data = Data(base64Encoded: payload) else {
            throw HpbError.invalid("invalid base64")
        }
        let raw = Array(data)
        guard raw.count >= 1 + 32 + 32 + 34, raw[0] == version else {
            throw HpbError.invalid("invalid payload")
        }
        let nonce = Array(raw[1..<33])
        let ciphertext = Array(raw[33..<raw.count - 32])
        let mac = Array(raw[(raw.count - 32)...])
        let keys = messageKeys(conversationKey, nonce)
        guard hmacSha256(keys.hmacKey, [nonce, ciphertext]) == mac else {
            throw HpbError.invalid("invalid MAC")
        }
        let plain = try unpad(ChaCha20.xor(key: keys.chachaKey, nonce12: keys.chachaNonce, data: ciphertext))
        guard let text = String(bytes: plain, encoding: .utf8) else {
            throw HpbError.invalid("invalid utf8")
        }
        return text
    }
}
